src/QsCompiler/Tests.Compiler/LambdaLiftingTests.fs

```diff
@@ -0,0 +1,733 @@
+        )
+
+        assertLambdaFunctorsByLine result lines.[0] "FooAdj" [ QsFunctor.Adjoint ]
+
+        let FooCtl =
+            TestUtils.getCallableWithName result Signatures.LambdaLiftingNS "FooCtl"
+            |> TestUtils.getBodyFromCallable
+
+        let lines = FooCtl |> TestUtils.getLinesFromSpecialization
+
+        Assert.True(
+            1 = Seq.length lines,
+            sprintf "Callable %O(%A) did not have the expected number of statements." FooCtl.Parent FooCtl.Kind
+        )
+
+        assertLambdaFunctorsByLine result lines.[0] "FooCtl" [ QsFunctor.Controlled ]
+
+        let FooAdjCtl =
+            TestUtils.getCallableWithName result Signatures.LambdaLiftingNS "FooAdjCtl"
+            |> TestUtils.getBodyFromCallable
+
+        let lines = FooAdjCtl |> TestUtils.getLinesFromSpecialization
+
+        Assert.True(
+            1 = Seq.length lines,
+            sprintf "Callable %O(%A) did not have the expected number of statements." FooAdjCtl.Parent FooAdjCtl.Kind
+        )
+
+        assertLambdaFunctorsByLine result lines.[0] "FooAdjCtl" [ QsFunctor.Adjoint; QsFunctor.Controlled ]
+
+    [<Fact>]
+    [<Trait("Category", "Parameters")>]
+    member this.``With Missing Params``() = compileLambdaLiftingTest 18
+
+    [<Fact>]
+    [<Trait("Category", "Return Values")>]
+    member this.``Use Parameter Single``() =
+        let result = compileLambdaLiftingTest 19
+
+        let original =
+            TestUtils.getCallableWithName result Signatures.LambdaLiftingNS "Foo"
+            |> TestUtils.getBodyFromCallable
+
+        let generated =
+            TestUtils.getCallablesWithSuffix result Signatures.LambdaLiftingNS "_Foo"
+            |> Seq.exactlyOne
+            |> TestUtils.getBodyFromCallable
+
+        let lines = generated |> TestUtils.getLinesFromSpecialization
+
+        let expectedContent = [| "return x;" |]
+        Assert.True((lines = expectedContent), "The generated callable did not have the expected content.")
+
+        let lines = original |> TestUtils.getLinesFromSpecialization
+
+        let expectedContent = [| sprintf "let lambda = %O(_);" generated.Parent; "let result = lambda(0);" |]
+        Assert.True((lines = expectedContent), "The original callable did not have the expected content.")
+
+    [<Fact>]
+    [<Trait("Category", "Return Values")>]
+    member this.``Use Parameter Tuple``() =
+        let result = compileLambdaLiftingTest 20
+
+        let original =
+            TestUtils.getCallableWithName result Signatures.LambdaLiftingNS "Foo"
+            |> TestUtils.getBodyFromCallable
+
+        let generated =
+            TestUtils.getCallablesWithSuffix result Signatures.LambdaLiftingNS "_Foo"
+            |> Seq.exactlyOne
+            |> TestUtils.getBodyFromCallable
+
+        let lines = generated |> TestUtils.getLinesFromSpecialization
+
+        let expectedContent = [| "return (y, x);" |]
```

Comment:
For these tests (Use Parameter Tuple, Use Parameter Tuple, Use Parameter and Closure, Use Parameter with Missing Params), I was hoping to have assertions about the order of names in the lifted function's parameter tuple. Otherwise, we don't know that `y` and `x` are referring to the correct parameters. (The idea being to catch bugs in how the lifter copies the lambda parameters and combines them with the closure parameters.)

Reply:
The tests `With Lots of Params` and `Use Closure With Params` already test that the parameters for lambdas are generated correctly in a variety of cases. I don't see how that applies to these tests you mention as these are for testing cases where the parameters are used, and don't in general have a variety of parameter configurations to test against.

Reply:
"With Lots of Params" and "Use Closure With Params" don't make assertions about the content of the lambdas. Because the assertions about the parameter tuples use a different test case from the assertions about the contents, the tests don't establish that a lambda is lifted with both the right variable names in the body and the right parameter names in the declaration.

Reply:
It seems unreasonable to have unit tests for cross products of independent components of a feature. We have tests for checking that parameter lists are generated correctly, and we have tests that check that the content is generated correctly. We even have tests for check that the parameters are generated correctly when those parameters are expected to be influenced by the content. Beyond what we have, I don't see a need to further complicate unit tests that already accomplish what they are meant to check, and I don't see a gap in test coverage.

Reply:
No, I agree, I didn't mean that there needs to be a cross product of testing properties. I had a simple additional test case in mind, but I think on second thought the current tests probably cover it.